### PR TITLE
Add `Proposer`, `UnsafeBlockSigner`, `BatchSubmitter` to tracked roles (works with and without Fault Proofs)

### PR DESCRIPTION
**Pull request by geoknee**
Towards https://github.com/ethereum-optimism/client-pod/issues/913
Closes https://github.com/ethereum-optimism/superchain-registry/issues/222

Changes: 
* Adds the above fields to the appropriate data structures, backfill missing data for all chains
* We still track `Proposer` for fault proof enabled chains (Context from slack https://oplabs-pbc.slack.com/archives/C073P71SGP2/p1718387014727779), but we get the information from a different contract on chain in that case
* Adds an e2e test case for `add-chain`, checking that it can add a fault proof enabled chain (covering all addresses, including these newly added ones)

### CODE REVIEW

File: add-chain/main.go [hunk 1]
```diff
@@ -176,12 +188,28 @@ func entrypoint(ctx *cli.Context) error {
...
+
+	version, err = strconv.Unquote(version)
+	if err != nil {
+		return fmt.Errorf("failed to parse OptimismPortalProxy.version(): %w", err)
+	}
+	majorVersion, err := strconv.ParseInt(strings.Split(version, ".")[0], 10, 32)
+	if err != nil {
+		return fmt.Errorf("failed to parse OptimismPortalProxy.version(): %w", err)
+	}
+	isFPAC := majorVersion >= 3
```
Comment (geoknee) [inline add-chain/main.go:205]: I suspect we want to just call this "isFaultProof", so I filed this https://github.com/ethereum-optimism/superchain-registry/issues/301 .

**Review by Inphi — COMMENTED** [review]
Found a few typos, but otherwise looks good.

File: add-chain/read_addresses.go [hunk 2]
```diff
@@ -31,49 +31,101 @@ var (
...
+	}
+	addresses[BatchSubmitter] = "0x" + hash[24:63]
+
+	if isFPAC {
+		// Proposer
+		address, err = castCall(addresses[PermissionedDisputeGame], "proposer()(address)", l1RpcUrl)
+		if err != nil {
+			return fmt.Errorf("could not retrieve address for Proposer")
+		}
+		addresses[UnsafeBlockSigner] = address
```
[the remark this answers is not in the document]
Comment (geoknee) [reply add-chain/read_addresses.go]: [dda403e](https://github.com/ethereum-optimism/superchain-registry/pull/298/commits/dda403e02ab5e2b8b8e2e61183ac65da08b896e1)

File: add-chain/read_addresses.go [hunk 3]
```diff
@@ -31,49 +31,101 @@ var (
...
+			return fmt.Errorf("could not retrieve address for Challenger")
+		}
+		addresses[Challenger] = address
+	} else {
+		// Proposer
+		address, err = castCall(addresses[L2OutputOracleProxy], "PROPOSER()(address)", l1RpcUrl)
+		if err != nil {
+			return fmt.Errorf("could not retrieve address for Proposer")
+		}
+		addresses[UnsafeBlockSigner] = address
```
[the remark this answers is not in the document]
Comment (geoknee) [reply add-chain/read_addresses.go]: [dda403e](https://github.com/ethereum-optimism/superchain-registry/pull/298/commits/dda403e02ab5e2b8b8e2e61183ac65da08b896e1)

**Review by Inphi — APPROVED** [review]

**Pull request merged**